components/script/dom/navigator.rs

```diff
@@ -128,15 +131,26 @@ impl NavigatorMethods for Navigator {
         self.vr.or_init(|| VR::new(&self.global()))
     }
 
+    // https://www.w3.org/TR/gamepad/#navigator-interface-extension
+    fn GetGamepads(&self) -> Root<GamepadList> {
+        let root = self.gamepads.or_init(|| {
+            GamepadList::new(&self.global(), Vec::new())
+        });
+
+        let vr_gamepads = self.Vr().get_gamepads();
+        root.sync(&vr_gamepads);
+        // TODO: Add not VR related gamepads
+        root
+    }
     // https://w3c.github.io/permissions/#navigator-and-workernavigator-extension
     fn Permissions(&self) -> Root<Permissions> {
         self.permissions.or_init(|| Permissions::new(&self.global()))
     }
 }
 
 impl Navigator {
-     pub fn handle_webvr_event(&self, event: WebVREventMsg) {
+     pub fn handle_webvr_events(&self, events: Vec<WebVREventMsg>) {
```

Comment:
We shouldn't need to go through the `Navigator` object just to handle `WebVR` events, should we? Please file a follow-up issue about this.